store/postgres/src/chain_store.rs

```diff
@@ -1813,6 +1841,120 @@ impl ChainStoreTrait for ChainStore {
     }
 }
 
+mod recent_blocks_cache {
+    use super::*;
+
+    type OffsetFromChainHead = BlockNumber;
+
+    struct Inner {
+        blocks: HashMap<OffsetFromChainHead, (BlockPtr, Option<json::Value>)>,
+        capacity: usize,
+    }
+
+    /// We cache the most recent blocks in memory to avoid overloading the
+    /// database with unnecessary queries close to the chain head. We invalidate
+    /// blocks:
+    /// - After 5 seconds.
+    /// - Whenever the chain head advances.
+    pub struct RecentBlocksCache {
+        // We protect everything with a global `RwLock` to avoid data races. Ugly...
+        inner: RwLock<Inner>,
+    }
+
+    impl RecentBlocksCache {
+        pub fn new(capacity: usize) -> Self {
+            RecentBlocksCache {
+                inner: RwLock::new(Inner {
+                    blocks: HashMap::new(),
+                    capacity,
+                }),
+            }
+        }
+
+        pub fn set_chain_head(&self, chain_head: BlockPtr) {
+            // We invalidate the entire cache when the chain head changes.
+            let mut inner = self.inner.write();
+            inner.blocks.clear();
+            inner.blocks.insert(0, (chain_head, None));
+        }
+
+        pub fn chain_head_ptr(&self) -> Option<BlockPtr> {
+            self.inner.read().blocks.get(&0).map(|p| p.0.clone())
+        }
+
+        pub fn clear(&self) {
+            self.inner.write().blocks.clear();
+        }
+
+        fn chain_head_number_inner(inner: &Inner) -> Option<BlockNumber> {
+            inner.blocks.get(&0).map(|p| p.0.number)
+        }
+
+        pub fn get_block(
+            &self,
+            ancestor: &BlockPtr,
+            offset: BlockNumber,
+        ) -> Option<(BlockPtr, Option<json::Value>)> {
+            Self::get_block_inner(&*self.inner.read(), ancestor, offset)
+        }
+
+        fn get_block_inner(
+            inner: &Inner,
+            ancestor: &BlockPtr,
+            offset: BlockNumber,
+        ) -> Option<(BlockPtr, Option<json::Value>)> {
+            // Check if the ancestor itself is in the cache.
+            let ancestor_offset_from_head = Self::chain_head_number_inner(inner)? - ancestor.number;
+            match inner
+                .blocks
+                .get(&ancestor_offset_from_head)
+                .map(|p| p.0.clone())
```

Comment:
It's minor, but is this `clone()` necessary?

Reply:
It is, unfortunately. We can't get a reference out of `self` because it's all guarded by the `RwLock`.